tests/components/humidifier/test_reproduce_state.py

```diff
@@ -0,0 +1,237 @@
+    assert turn_on_calls[0].data == {"entity_id": ENTITY_1}
+
+    assert len(turn_off_calls) == 1
+    assert turn_off_calls[0].domain == "humidifier"
+    assert turn_off_calls[0].data == {"entity_id": ENTITY_2}
+
+    # Make sure we didn't call services for missing attributes
+    assert len(mode_calls) == 0
+    assert len(humidity_calls) == 0
+
+
+async def test_multiple_attrs(hass):
+    """Test turn on with multiple attributes."""
+    hass.states.async_set(ENTITY_1, STATE_OFF, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    await async_reproduce_states(
+        hass, [State(ENTITY_1, STATE_ON, {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45})]
+    )
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 1
+    assert turn_on_calls[0].data == {"entity_id": ENTITY_1}
+    assert len(turn_off_calls) == 0
+    assert len(mode_calls) == 1
+    assert mode_calls[0].data == {"entity_id": ENTITY_1, "mode": "normal"}
+    assert len(humidity_calls) == 1
+    assert humidity_calls[0].data == {"entity_id": ENTITY_1, "humidity": 45}
+
+
+async def test_turn_off_multiple_attrs(hass):
+    """Test set mode and humidity for off state."""
+    hass.states.async_set(ENTITY_1, STATE_ON, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    await async_reproduce_states(
+        hass, [State(ENTITY_1, STATE_OFF, {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 45})]
+    )
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 0
+    assert len(turn_off_calls) == 1
+    assert turn_off_calls[0].data == {"entity_id": ENTITY_1}
+    assert len(mode_calls) == 0
+    assert len(humidity_calls) == 0
+
+
+async def test_multiple_modes(hass):
+    """Test that multiple states gets calls."""
+    hass.states.async_set(ENTITY_1, STATE_OFF, {})
+    hass.states.async_set(ENTITY_2, STATE_OFF, {})
+
+    turn_on_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_ON)
+    turn_off_calls = async_mock_service(hass, DOMAIN, SERVICE_TURN_OFF)
+    mode_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_MODE)
+    humidity_calls = async_mock_service(hass, DOMAIN, SERVICE_SET_HUMIDITY)
+
+    await async_reproduce_states(
+        hass,
+        [
+            State(ENTITY_1, STATE_ON, {ATTR_MODE: MODE_ECO, ATTR_HUMIDITY: 40}),
+            State(ENTITY_2, STATE_ON, {ATTR_MODE: MODE_NORMAL, ATTR_HUMIDITY: 50}),
+        ],
+    )
+
+    await hass.async_block_till_done()
+
+    assert len(turn_on_calls) == 2
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5745133)